src/Realtime.ts

```diff
@@ -203,22 +237,133 @@ abstract class BaseGameState {
   }
 
   abstract applyDiffs(diffs: BoardDiff[]): void;
+  abstract getBoardState(): LocalBoardState;
+
+  // high level APIs:
+  setNumber(
+    squareIdx: number,
+    digit: number | null,
+    options?: { automaticallyRemoveMarks?: boolean }
+  ): void {
+    options = { automaticallyRemoveMarks: false, ...(options ?? {}) };
+    // HACK: We shouldn't read the board state here. There's no risk of this
+    // behavior causing a desync, but this could still result in unexpected
+    // results.
+    if (nullthrows(this.getBoardState().squares.get(squareIdx)).get("locked")) {
+      return;
+    }
```

Comment:
The server and diff system understand that locked squares shouldn't get updated, but this can generate diffs that change "affected" squares, which the diff system doesn't understand.

I think I can solve this by adding a simple "BreakIfLocked" instruction to the diff system, introducing a small amount of control flow.

For now this is fine because the UI doesn't expose an easy way to toggle the locked state of a square, so the chance that the client's view of the board is out-of-date enough for this to matter is low.